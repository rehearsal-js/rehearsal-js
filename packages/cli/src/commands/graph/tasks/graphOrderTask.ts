import { dirname, resolve } from 'node:path';
import { Worker } from 'node:worker_threads';
import { writeFile } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';
import debug from 'debug';
import { ListrDefaultRenderer, ListrTask } from 'listr2';
import type { PackageEntry, GraphCommandContext, GraphTaskOptions } from '../../../types.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const workerPath = resolve(__dirname, 'graphWorker.js');

const DEBUG_CALLBACK = debug('rehearsal:cli:graphOrderTask');

export function graphOrderTask(
  options: GraphTaskOptions,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _ctx?: GraphCommandContext
): ListrTask<GraphCommandContext, ListrDefaultRenderer> {
  return {
    title: 'Analyzing project dependency graph',
    options: { persistentOutput: true },
    async task(ctx: GraphCommandContext, task) {
      let order: { packages: PackageEntry[] };
      const { srcDir, output, basePath } = options;
      let selectedPackageName: string;

      if (process.env['TEST'] === 'true') {
        // Do this on the main thread because there are issues with resolving worker scripts for worker_threads in vitest
        const { intoGraphOutput } = await import('./graphWorker.js').then((m) => m);
        const { getMigrationOrder } = await import('@rehearsal/migration-graph').then((m) => m);
        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-unsafe-call
        order = intoGraphOutput(getMigrationOrder(srcDir, { basePath }), srcDir);
      } else {
        order = await new Promise<{ packages: PackageEntry[] }>((resolve, reject) => {
          task.title = 'Analyzing project dependency graph ...';

          // Run graph traversal in a worker thread so the ui thread doesn't hang
          const worker = new Worker(workerPath, {
            workerData: JSON.stringify({ srcDir, basePath: options.basePath }),
          });

          worker.on('message', (packages: { packages: PackageEntry[] }) => {
            console.log(packages);
            resolve(packages);
          });

          worker.on('error', reject);
          worker.on('exit', (code) => {
            if (code !== 0) {
              reject(new Error(`Worker stopped with exit code ${code}`));
            }
          });
        });
      }

      DEBUG_CALLBACK(`order: ${JSON.stringify(order, null, 2)}`);
      DEBUG_CALLBACK(`ctx: ${JSON.stringify(ctx, null, 2)}`);

      // if explicit child package is passed in use that
      if (ctx?.childPackage) {
        ctx.jsSourcesAbs = getGraphFilesAbs(srcDir, ctx.childPackage, order.packages[0].files);

        return;
      }

      // dont prompt just write the file
      if (output) {
        task.title = `Writing graph to ${output} ...`;

        await writeFile(output, JSON.stringify(order, null, 2));
      } else {
        // if more than 1 package found prompt the user to select one
        if (order.packages.length > 1) {
          selectedPackageName = await task.prompt<string>([
            {
              type: 'Select',
              name: 'packageSelection',
              message:
                'We found the following packages. Select a packages to see the order of files:',
              choices: order.packages.map((p) => p.name),
            },
          ]);
        } else {
          selectedPackageName = order.packages[0].name;
        }

        const entry = getPackageEntry(order, selectedPackageName);

        task.output = `Graph order for '${selectedPackageName}':\n\n${entry}`;
      }
    },
  };
}

// get all the js | gjs files from the graph and return as absolute paths. the graph will filter the extensions
function getGraphFilesAbs(basePath: string, childPackage: string, files: string[]): string[] {
  // filter out files that are not in the child package
  files = files.filter((file) => file.startsWith(childPackage));

  return (
    files.map((file) => {
      // all files are relative to basePath
      return resolve(basePath, file);
    }) ?? []
  );
}

function getPackageEntry(
  order: { packages: PackageEntry[] },
  selectedPackageName: string
): string | undefined {
  return order.packages.find((pkg) => pkg.name === selectedPackageName)?.files.join('\n');
}
