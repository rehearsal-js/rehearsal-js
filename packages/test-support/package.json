{
  "name": "@rehearsal/test-support",
  "version": "3.0.1-beta",
  "description": "Test support utilities for rehearsal",
  "license": "BSD-2-Clause",
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/src/index.d.ts",
      "import": "./dist/src/index.js"
    },
    "./*": "./*"
  },
  "main": "dist/src/index.js",
  "types": "dist/src/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc -b",
    "fixtures:clean": "rm -rf ./fixtures/ember/*/node_modules",
    "fixtures:prepare": "node ./scripts/setup-fixtures.js ./fixtures/ember/app-template ./fixtures/ember/addon-template ./fixtures/ember/app-ts-template",
    "postinstall": "pnpm fixtures:prepare",
    "lint": "npm-run-all lint:*",
    "lint:tsc-src": "tsc --noEmit",
    "lint:tsc-test": "tsc --noEmit --project test/tsconfig.json",
    "test": "vitest --run --config ./vitest.config.ts --coverage",
    "test:slow": "vitest --run --config ./vitest.config.slow.ts",
    "version": "pnpm version"
  },
  "dependencies": {
    "find-up": "^6.3.0",
    "fixturify": "^3.0.0",
    "fixturify-project": "^5.2.0",
    "rimraf": "^4.4.1",
    "scenario-tester": "^2.1.2",
    "tmp": "^0.2.1"
  },
  "devDependencies": {
    "@vitest/coverage-c8": "^0.30.1",
    "ember-cli": "^4.11.0",
    "ember-data": "^4.11.3",
    "ember-source": "^4.12.0",
    "vitest": "^0.30.0"
  },
  "peerDependencies": {
    "typescript": "^5.0"
  },
  "packageManager": "pnpm@8.2.0",
  "volta": {
    "extends": "../../package.json"
  }
}
