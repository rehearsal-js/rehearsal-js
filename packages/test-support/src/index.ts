export * from './files.js';
export * from './library.js';
export * from './project.js';
export * from './scenarios.js';
