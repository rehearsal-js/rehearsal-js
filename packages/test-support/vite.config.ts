import { defineConfig } from 'vitest/config';

//eslint-disable-next-line import/no-default-export
export default defineConfig({
  test: {
    testTimeout: 100_000,
    hookTimeout: 100_000,
    watchExclude: ['package.json', '**/fixtures/**'],
    threads: false,
  },
});
