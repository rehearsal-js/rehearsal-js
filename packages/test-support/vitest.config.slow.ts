import { defineConfig } from 'vitest/config';


// explicit includes for slow test otherwise it'll run in `test`

//eslint-disable-next-line import/no-default-export
export default defineConfig({
  test: {
    testTimeout: 500_000,
    hookTimeout: 50_000,
    watchExclude: ['package.json', '**/fixtures/**'],
    include: ['./test/scenarios-app.test.ts']
  },
});
