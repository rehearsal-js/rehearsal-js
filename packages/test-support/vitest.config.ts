import { defineConfig } from 'vitest/config';

// we dont want to run these slow tests in this config
// these will all run by default in `test:slow` instead which is run in CI
const exclude = ['./test/scenarios-app.test.ts'];

//eslint-disable-next-line import/no-default-export
export default defineConfig({
  test: {
    testTimeout: 500_000,
    hookTimeout: 50_000,
    watchExclude: ['package.json', '**/fixtures/**'],
    sequence: {
      shuffle: true,
    },
    include: ['./test/**/*.test.ts'],
    exclude,
  },
});
