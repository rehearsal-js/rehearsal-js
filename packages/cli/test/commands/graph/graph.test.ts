import { join } from 'node:path';
import { existsSync, readFileSync } from 'node:fs';
import { readFile } from 'node:fs/promises';
import { Readable } from 'node:stream';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { createLogger, format, transports } from 'winston';

import { getEmberProject } from '@rehearsal/test-support';
import {
  listrTaskRunner,
  KEYS,
  cleanOutput,
  createOutputStream,
  prepareProject,
  sendKey,
} from '../../test-helpers/index.js';
import { graphOrderTask } from '../../../src/commands/graph/tasks/graphOrderTask.js';
import type { PackageEntry, GraphCommandContext } from '../../../src/types.js';
import type { DirJSON } from 'fixturify';
import type { Project } from 'fixturify-project';

const logger = createLogger({
  transports: [new transports.Console({ format: format.cli() })],
});

describe('Task: graphOrderTask', () => {
  let project: Project;
  let outputStream: Readable;
  let output = '';
  vi.spyOn(console, 'info').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
  });
  vi.spyOn(console, 'log').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
  });
  vi.spyOn(console, 'error').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
  });
  vi.spyOn(logger, 'warn').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
    return logger;
  });
  vi.spyOn(logger, 'error').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
    return logger;
  });

  beforeEach(async () => {
    output = '';
    project = prepareProject('multi_packages');
    outputStream = createOutputStream();
    // tests below assume creation
    delete project.files['tsconfig.json'];
    await project.write();
  });

  afterEach(() => {
    output = '';
    vi.clearAllMocks();
    project.dispose();
  });

  test('can output a graph.json file', async () => {
    const options = {
      srcDir: project.baseDir,
      basePath: project.baseDir,
      output: join(project.baseDir, 'graph.json'),
      devDeps: false,
      deps: true,
      ignore: [],
    };

    await listrTaskRunner<GraphCommandContext>([graphOrderTask(options)]);

    expect(existsSync(options.output)).toBe(true);

    const graph = JSON.parse(await readFile(options.output, 'utf-8')) as {
      packages: PackageEntry[];
    };

    expect(graph.packages.length).toBe(1);
    expect(graph.packages[0].files).toMatchObject([
      'index.js',
      'module-a/index.js',
      'module-a/src/baz.js',
      'module-a/src/foo.js',
      'module-b/index.js',
      'module-b/src/tires.js',
      'module-b/src/car.js',
      'src/foo/baz.js',
      'src/foo/biz.js',
      'src/foo/buz/biz.js',
      'src/index.js',
    ]);
  });

  test('can output to stdout for sub-package', async () => {
    const options = {
      srcDir: `${project.baseDir}/module-a`,
      basePath: project.baseDir,
      output: join(project.baseDir, 'graph.json'),
      devDeps: false,
      deps: true,
      ignore: [],
    };

    await listrTaskRunner<GraphCommandContext>([graphOrderTask(options)]);

    expect(existsSync(options.output)).toBe(true);

    const graph = JSON.parse(await readFile(options.output, 'utf-8')) as {
      packages: PackageEntry[];
    };

    expect(graph.packages.length).toBe(1);
    expect(graph.packages[0].files).toMatchObject([
      'module-a/index.js',
      'module-a/src/baz.js',
      'module-a/src/foo.js',
    ]);
  });

  test('can print graph order to stdout', async () => {
    const options = {
      srcDir: project.baseDir,
      basePath: project.baseDir,
      devDeps: false,
      deps: true,
      ignore: [],
    };

    await listrTaskRunner<GraphCommandContext>([graphOrderTask(options)]);

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test('can print graph order to stdout in an ember app with in-repo addons', async () => {
    project = getEmberProject('app-with-in-repo-addon');

    await project.write();

    const options = {
      srcDir: project.baseDir,
      basePath: project.baseDir,
      devDeps: false,
      deps: true,
      ignore: [],
    };

    outputStream.on('data', (line: string) => {
      if (line.includes('We found the following packages.')) {
        sendKey(KEYS.ENTER);
      }
    });

    await listrTaskRunner<GraphCommandContext>([graphOrderTask(options)]);

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });

  test('subsets ember graph when entered via addon', async () => {
    project = addWorkspaces(getEmberProject('app-with-in-repo-addon'));

    project.mergeFiles(someOtherAddons(project));

    await project.write();

    const options = {
      srcDir: project.baseDir + '/lib/some-other-addon',
      basePath: project.baseDir,
      output: join(project.baseDir, 'graph.json'),
      devDeps: false,
      deps: true,
      ignore: [],
    };

    await listrTaskRunner<GraphCommandContext>([graphOrderTask(options)]);

    expect(existsSync(options.output)).toBe(true);

    const graph = JSON.parse(await readFile(options.output, 'utf-8')) as {
      packages: PackageEntry[];
    };

    expect(graph.packages.length).toBe(2);

    expect(graph.packages[0].files).toMatchObject([
      'lib/some-addon/addon/utils/thing.js',
      'lib/some-addon/addon/components/greet.js',
    ]);

    expect(graph.packages[1].files).toMatchObject(['lib/some-other-addon/addon/index.js']);
  });

  test('only follows devDependecies when explicitly asked', async () => {
    project = addWorkspaces(getEmberProject('app-with-in-repo-addon'));

    project.mergeFiles(someOtherAddons(project));

    await project.write();

    const options = {
      srcDir: project.baseDir + '/lib/some-other-addon',
      basePath: project.baseDir,
      output: join(project.baseDir, 'graph.json'),
      devDeps: false,
      deps: true,
      ignore: [],
    };

    await listrTaskRunner<GraphCommandContext>([graphOrderTask(options)]);

    expect(existsSync(options.output)).toBe(true);

    const graph = JSON.parse(await readFile(options.output, 'utf-8')) as {
      packages: PackageEntry[];
    };

    expect(graph.packages.length).toBe(2);

    expect(graph.packages[0].files).toMatchObject([
      'lib/some-addon/addon/utils/thing.js',
      'lib/some-addon/addon/components/greet.js',
    ]);

    expect(graph.packages[1].files).toMatchObject(['lib/some-other-addon/addon/index.js']);

    const withDevDeps = {
      srcDir: project.baseDir + '/lib/some-other-addon',
      basePath: project.baseDir,
      devDeps: true,
      deps: true,
      ignore: [],
      output: join(project.baseDir, 'graph.json'),
    };

    await listrTaskRunner<GraphCommandContext>([graphOrderTask(withDevDeps)]);

    expect(existsSync(withDevDeps.output)).toBe(true);

    const graphWithDevDeps = JSON.parse(await readFile(withDevDeps.output, 'utf-8')) as {
      packages: PackageEntry[];
    };

    expect(graphWithDevDeps.packages.length).toBe(3);

    expect(graphWithDevDeps.packages[0].files).toMatchObject([
      'lib/some-addon/addon/utils/thing.js',
      'lib/some-addon/addon/components/greet.js',
    ]);

    expect(graphWithDevDeps.packages[1].files).toMatchObject([
      'lib/some-test-package/addon-test-support/index.js',
    ]);

    expect(graphWithDevDeps.packages[2].files).toMatchObject([
      'lib/some-other-addon/addon/index.js',
    ]);
  });

  test('can ignore packages from graph', async () => {
    project = addWorkspaces(getEmberProject('app-with-in-repo-addon'));

    project.mergeFiles(someOtherAddons(project));

    await project.write();

    const options = {
      srcDir: project.baseDir + '/lib/some-other-addon',
      basePath: project.baseDir,
      output: join(project.baseDir, 'graph.json'),
      devDeps: true,
      deps: true,
      ignore: ['some-test-package'],
    };

    await listrTaskRunner<GraphCommandContext>([graphOrderTask(options)]);

    expect(existsSync(options.output)).toBe(true);

    const graph = JSON.parse(await readFile(options.output, 'utf-8')) as {
      packages: PackageEntry[];
    };

    expect(graph.packages.length).toBe(2);

    expect(graph.packages[0].files).toMatchObject([
      'lib/some-addon/addon/utils/thing.js',
      'lib/some-addon/addon/components/greet.js',
    ]);

    expect(graph.packages[1].files).toMatchObject(['lib/some-other-addon/addon/index.js']);
  });

  test('can print graph order to a file for an ember app with in-repo addons', async () => {
    project = getEmberProject('app-with-in-repo-addon');

    await project.write();

    const options = {
      basePath: project.baseDir,
      srcDir: project.baseDir,
      output: join(project.baseDir, 'graph.json'),
      devDeps: false,
      deps: true,
      ignore: [],
    };

    outputStream.on('data', (line: string) => {
      if (line.includes('We found the following packages.')) {
        sendKey(KEYS.ENTER);
      }
    });

    await listrTaskRunner<GraphCommandContext>([graphOrderTask(options)]);

    expect(existsSync(options.output)).toBe(true);

    const graph = JSON.parse(await readFile(options.output, 'utf-8')) as {
      packages: PackageEntry[];
    };

    expect(graph.packages.length).toBe(2);
    expect(graph.packages[0].files).toMatchObject([
      'lib/some-addon/addon/utils/thing.js',
      'lib/some-addon/addon/components/greet.js',
    ]);

    expect(graph.packages[1].files).toMatchObject([
      'app/app.js',
      'app/services/locale.js',
      'app/components/salutation.js',
      'app/router.js',
      'tests/acceptance/index-test.js',
      'tests/test-helper.js',
      'tests/unit/services/locale-test.js',
    ]);
  });

  test('can print graph order to a file for an ember app with in-repo addons when the module name and package name differ', async () => {
    project = getEmberProject('app-with-in-repo-addon', true);

    await project.write();

    expect(
      readFileSync(join(project.baseDir, 'lib', 'some-addon', 'index.js'), 'utf-8').includes(
        'name: "some-addon"'
      )
    ).toBe(true);

    expect(
      readFileSync(join(project.baseDir, 'lib', 'some-addon', 'package.json'), 'utf-8').includes(
        '"name": "@company/some-addon"'
      )
    ).toBe(true);

    const options = {
      srcDir: project.baseDir,
      basePath: project.baseDir,
      output: join(project.baseDir, 'graph.json'),
      devDeps: false,
      deps: true,
      ignore: [],
    };

    outputStream.on('data', (line: string) => {
      if (line.includes('We found the following packages.')) {
        sendKey(KEYS.ENTER);
      }
    });

    await listrTaskRunner<GraphCommandContext>([graphOrderTask(options)]);

    expect(existsSync(options.output)).toBe(true);

    const graph = JSON.parse(await readFile(options.output, 'utf-8')) as {
      packages: PackageEntry[];
    };

    expect(graph.packages.length).toBe(2);
    expect(graph.packages[0].files).toMatchObject([
      'lib/some-addon/addon/utils/thing.js',
      'lib/some-addon/addon/components/greet.js',
    ]);

    expect(graph.packages[1].files).toMatchObject([
      'app/app.js',
      'app/services/locale.js',
      'app/components/salutation.js',
      'app/router.js',
      'tests/acceptance/index-test.js',
      'tests/test-helper.js',
      'tests/unit/services/locale-test.js',
    ]);
  });
});

function someOtherAddons(project: Project): DirJSON {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
  project.pkg['ember-addon'].paths.push('lib/some-other-addon');
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
  project.pkg['ember-addon'].paths.push('lib/some-test-package');
  return {
    lib: {
      'some-test-package': {
        'addon-test-support': {
          'index.js': 'export const thing = "thing"',
        },
        'index.js': 'module.exports = { name: "some-test-package" }',
        'package.json': JSON.stringify({ name: 'some-test-package', keywords: ['ember-addon'] }),
      },
      'some-other-addon': {
        addon: {
          'index.js': 'import Greet from "some-addon/components/greet"',
        },
        'index.js': 'module.exports = { name: "some-other-addon" }',
        'package.json': JSON.stringify({
          name: '@company/some-other-addon',
          dependencies: { 'some-addon': '*' },
          devDependencies: { 'some-test-package': '*' },
          keywords: ['ember-addon'],
        }),
      },
    },
  };
}

function addWorkspaces(project: Project): Project {
  project.pkg.workspaces = ['./lib/*'];
  return project;
}
