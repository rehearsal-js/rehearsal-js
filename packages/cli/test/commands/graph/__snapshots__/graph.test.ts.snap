// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Task: graphOrderTask > can print graph order to stdout 1`] = `
"[STARTED] Analyzing project dependency graph
[DATA] Graph order for .:
[DATA] index.js
[DATA] module-a/index.js
[DATA] module-a/src/baz.js
[DATA] module-a/src/foo.js
[DATA] module-b/index.js
[DATA] module-b/src/tires.js
[DATA] module-b/src/car.js
[DATA] src/bizz.ts
[DATA] src/foo/baz.js
[DATA] src/foo/biz.js
[DATA] src/foo/buz/biz.js
[DATA] src/index.js
[SUCCESS] Analyzing project dependency graph"
`;

exports[`Task: graphOrderTask > can print graph order to stdout in an ember app with in-repo addons 1`] = `
"[STARTED] Analyzing project dependency graph
[DATA] ? We found the following packages. Select a packages to see the order of files:  
[DATA]  lib/some-addon
[DATA]   .
[DATA]  We found the following packages. Select a packages to see the order of files:  lib/some-addon
[DATA] Graph order for lib/some-addon:
[DATA] lib/some-addon/addon/utils/thing.js
[DATA] lib/some-addon/addon/components/greet.js
[SUCCESS] Analyzing project dependency graph"
`;
